File: structural/typescript/composite_structured/leaves.ts
import { FileSystemComponent, Employee, Graphic } from './components';

export class File extends FileSystemComponent {
  private name: string;
  private size: number;
  
  constructor(name: string, size: number) { super(); this.name = name; this.size = size; }
  display(indent: string = ''): void { console.log(`\${indent}File: \${this.name} (\${this.size} bytes)`); }
  getSize(): number { return this.size; }
}

export class Worker extends Employee {
  private name: string;
  private position: string;
  
  constructor(name: string, position: string) { super(); this.name = name; this.position = position; }
  print(indent: string = ''): void { console.log(`\${indent}Employee: \${this.name} - \${this.position}`); }
}

export class Circle extends Graphic {
  private name: string;
  constructor(name: string) { super(); this.name = name; }
  draw(indent: string = ''): void { console.log(`\${indent}Drawing Circle: \${this.name}`); }
}

export class Square extends Graphic {
  private name: string;
  constructor(name: string) { super(); this.name = name; }
  draw(indent: string = ''): void { console.log(`\${indent}Drawing Square: \${this.name}`); }
}

File: structural/dart/composite_structured/leaves.dart
import 'components.dart';

class FileComponent extends FileSystemComponent {
  final String name;
  final int size;
  FileComponent(this.name, this.size);
  
  @override
  void display([String indent = '']) { print('\$indent File: \$name (\$size bytes)'); }
  @override
  int getSize() { return size; }
}

class Worker extends Employee {
  final String name;
  final String position;
  Worker(this.name, this.position);
  
  @override
  void print([String indent = '']) { print('\$indent Employee: \$name - \$position'); }
}

class CircleGraphic extends Graphic {
  final String name;
  CircleGraphic(this.name);
  @override
  void draw([String indent = '']) { print('\$indent Drawing Circle: \$name'); }
}

class SquareGraphic extends Graphic {
  final String name;
  SquareGraphic(this.name);
  @override
  void draw([String indent = '']) { print('\$indent Drawing Square: \$name'); }
}
